src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -0,0 +1,518 @@
+        # the controller secret.  Pull-secret will be handled separately.
+        base64_data: Dict[str, str] = dict()
+        for name in secret_names:
+            secret: Secret = await self.read_secret(
+                name=name, namespace=source_ns
+            )
+            # Retrieve matching keys
+            for key in secret.data:
+                if key in secret_keys:
+                    base64_data[key] = secret.data[key]
+        # There's no point in decoding it; all we're gonna do is pass it
+        # down to create a secret as base64 anyway.
+        if "token" in base64_data:
+            raise RuntimeError("'token' must come from the user token")
+        if not token:
+            raise RuntimeError("User token cannot be empty")
+        base64_data["token"] = str(base64.b64encode(token.encode("utf-8")))
+        return base64_data
+
+    async def create_secret(
+        self,
+        name: str,
+        namespace: str,
+        data: Dict[str, str],
+        secret_type: str = "Opaque",
+        immutable: bool = True,
+    ) -> None:
+        secret = V1Secret(
+            data=data,
+            type=secret_type,
+            immutable=immutable,
+            metadata=self.get_std_metadata(name),
+        )
+        await self.api.create_namespaced_secret(namespace, secret)
+        return
+
+    async def read_secret(
+        self,
+        name: str,
+        namespace: str,
+    ) -> Secret:
+        secret: V1Secret = await self.api.read_namespaced_secret(
+            name, namespace
+        )
+        secret_type = secret.type
+        return Secret(data=secret.data, secret_type=secret_type)
+
+    async def create_configmap(
+        self,
+        name: str,
+        namespace: str,
+        data: Dict[str, str],
+        immutable: bool = True,
+    ) -> None:
+        configmap = V1ConfigMap(
+            data=data,
+            immutable=immutable,
+            metadata=self.get_std_metadata(name="configmap"),
+        )
+        await self.api.create_namespaced_configmap(namespace, configmap)
+
+    async def create_network_policy(
+        self,
+        name: str,
+        namespace: str,
+    ) -> None:
+        api = client.NetworkingV1Api(self.k8s_api)
+        policy = V1NetworkPolicy(
+            metadata=self.get_std_metadata(name),
+            spec=V1NetworkPolicySpec(
+                policy_types=["Ingress"],
+                pod_selector=V1LabelSelector(
+                    match_labels={"app": "jupyterhub", "component": "hub"}
+                ),
+                ingress=[
+                    V1NetworkPolicyIngressRule(
+                        ports=[
+                            V1NetworkPolicyPort(
+                                port={"port": "http"},
+                            ),
+                            V1NetworkPolicyPort(
+                                port={"port": 8081},
+                            ),
```

Comment:
We're hard-coding a bunch of stuff here about JupyterHub setup that's determined by Zero-to-JupyterHub. Probably no way to avoid that (and I don't think it should be configurable), but I would lift the match labels and port list into constants so that we can find them easily if Zero-to-JupyterHub changes something.